file/src/main/java/akka/stream/alpakka/file/javadsl/FileTailSource.java

```diff
@@ -43,12 +52,36 @@
    */
   public static Source<ByteString, NotUsed> create(
       Path path, int maxChunkSize, long startingPosition, java.time.Duration pollingInterval) {
-    return Source.fromGraph(
-        new akka.stream.alpakka.file.impl.FileTailSource(
-            path,
-            maxChunkSize,
-            startingPosition,
-            JavaDurationConverters.asFiniteDuration(pollingInterval)));
+    return Source
+        .fromGraph(
+            new akka.stream.alpakka.file.impl.FileTailSource(
+                path,
+                maxChunkSize,
+                startingPosition,
+                JavaDurationConverters.asFiniteDuration(pollingInterval))
+        )
+        .merge(fileCheckSource(pollingInterval, path), true);
+  }
+
+  /**
+   * Periodically checks if the file at `path` exists every `pollingInterval`. Shutdown the stream
+   * when the file is not found using an empty Source.
+   * @param pollingInterval Interval to check if file exists.
+   * @param path a file to check
+   */
+  private static Source<ByteString, Cancellable> fileCheckSource(java.time.Duration pollingInterval, Path path) {
+    return Source
+        .tick(pollingInterval, pollingInterval, new Tick())
+        .mapConcat(tick -> {
+          if (Files.exists(path)) {
+            return Collections.<ByteString>emptyList();
+          }
+          throw new FileNotFoundException();
+        })
+        .recoverWith(new PFBuilder<Throwable, Source<ByteString, NotUsed>>()
+            .match(FileNotFoundException.class, t -> Source.empty())
+            .build()
+        );
```

Comment:
If you are sure you want this as a separate stream + a public API I think `.takeWhile(e -> fileExists)` would be a less roundabout way of doing it.

However spontaneously it sounds like a concern that should be covered by the graph stage rather than through composition, now you will end up with a race condition where the last bytes the FileTailSource may or may not arrive before the ´fileCheckSource´ completes if the file is deleted.

Reply:
@johanandren Thanks for the review.  That's a good point about the race condition, I hadn't considered that.

The original implementation was in the `FileTailSource` implementation.  One of the issues we found was that checking if the file exists is a synchronous operation, and even if it were made asynchronous with a timer, it could still block the stage from processing other log content.  I had considered creating another actor within the stage to do the checking and message back if the file was detected as deleted, but a separate stream seemed a little more elegant at the time.

Reply:
Moved this implementation to a documented use case with the race condition caveat.